src/AutoHashEqualsCached.jl

```diff
@@ -2,339 +2,63 @@
-        push!(result.args, esc(:(function $base_hash_name(x::$type_name, h::UInt)
-            $defined_hash_name(x, h)
-            end)))
-        push!(result.args, esc(:(function $base_hash_name(x::$type_name)
-            $defined_hash_name(x)
-            end)))
-    end
-
-    # add function Base.show
-    push!(result.args, esc(:(function $Base._show_default(io::IO, x::$type_name)
-        $_show_default_auto_hash_equals_cached(io, x)
-        end)))
-
-    # Add functions to interoperate with Match if loaded
-    # at the time the macro is expanded.
-    if_has_package("Match", Base.UUID("7eb4fadd-790c-5f42-8a69-bfa0b872bfbf"), v"2") do pkg
-        if :match_fieldnames in names(pkg; all=true)
-            push!(result.args, esc(:(function $pkg.match_fieldnames(::Type{$type_name})
-                $((member_names...,))
-                end)))
-        end
-    end
-
-    equalty_impl = foldl(
-        (r, f) -> :($r && $isequal($getfield(a, $(QuoteNode(f))), $getfield(b, $(QuoteNode(f))))),
-        member_names;
-        init = :(a._cached_hash == b._cached_hash))
-
-    if isnothing(where_list)
-        # add == for non-generic types
-        push!(result.args, esc(quote
-            function $Base.:(==)(a::$type_name, b::$type_name)
-                $equalty_impl
-            end
-        end))
-    else
-        # We require the type be the same (including type arguments) for two instances to be equal
-        push!(result.args, esc(quote
-            function $Base.:(==)(a::$full_type_name, b::$full_type_name) where {$(where_list...)}
-                $equalty_impl
-            end
-        end))
-        # for generic types, we add an external constructor to perform ctor type inference:
-        push!(result.args, esc(quote
-            $type_name($(member_decls...)) where {$(where_list...)} = $full_type_name($(member_names...))
-        end))
-    end
-
-    return result
+    typ = args[end]
+    auto_hash_equals_impl(__source__, typ; kwargs...)
 end
 
 """
     @auto_hash_equals_cached struct Foo ... end
 
-Causes the struct to have an additional hidden field named `_cached_hash` that is
-computed and stored at the time of construction.  Produces constructors and specializes
-the behavior of `Base.show` to maintain the illusion that the field does not exist.
-Two different instantiations of a generic type are considered not equal.
-
-Also produces specializations of `Base.hash` and `Base.==`:
-
-- `Base.==` is implemented as an elementwise test for `isequal`.
-- `Base.hash` just returns the cached hash value.
+Shorthand for @auto_hash_equals cache=true struct Foo ... end
 """
-macro auto_hash_equals_cached(typ::Expr)
-    auto_hash_equals_cached_impl(__source__, nothing, typ)
-end
-macro auto_hash_equals_cached(alt_hash_name, typ::Expr)
-    auto_hash_equals_cached_impl(__source__, alt_hash_name, typ)
+macro auto_hash_equals_cached(typ)
+    esc(Expr(:macrocall, var"@auto_hash_equals", __source__, :(cache = true), typ))
```

Comment:
could this be written closer to source without the expr? or does that lose the `__source__` provenance?

Reply:
Yes, this is necessary to preserve `__source__`.